File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.cpp
class Solution {
public:
    int numSteps(string s) {
        string temp = s;
        int n = temp.size();

        int steps = 0;
        while (n > 1) {
            // cout << temp << endl;
            if (temp[n - 1] == '1') {
                temp = addOne(temp);
                steps++;
                // cout << "temp adding one: " << temp << endl; 
            }

            temp.pop_back();
            steps++;
            n = temp.size();
        }
        
        return steps;
    }

private:
    string addOne(string number) {
        int n = number.size();

        int carry = 0;
        string res = "";
        for (int i = n - 1; i >= 0; i--) {
            int d1 = number[i] - '0';
            int d2 = i == (n - 1) ? 1 : 0;
            int sum = 0;

            if (carry == 0) {
                if (d1 == 0 && d2 == 0) {
                    sum = 0;
                    carry = 0;
                } else if (d1 == 0 && d2 == 1) {
                    sum = 1;
                    carry = 0;
                } else if (d1 == 1 && d2 == 0) {
                    sum = 1;
                    carry = 0;
                } else if (d1 == 1 && d2 == 1) {
                    sum = 0;
                    carry = 1;
                }
            } else if (carry == 1) {
                if (d1 == 0 && d2 == 0) {
                    sum = 1;
                    carry = 0;
                } else if (d1 == 0 && d2 == 1) {
                    sum = 0;
                    carry = 1;
                } else if (d1 == 1 && d2 == 0) {
                    sum = 0;
                    carry = 1;
                } else if (d1 == 1 && d2 == 1) {
                    sum = 1;
                    carry = 1;
                }
            }

            res += (sum + '0');
        }

        if (carry == 1) {
            res += (carry + '0');
        }

        reverse(res.begin(), res.end());
        return res;
    }
};

File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.java
class Solution {
    public int numSteps(String s) {
        List<Character> charList = s.chars()
                .mapToObj(c -> (char) c)
                .collect(Collectors.toList());

        int ans = 0;
        while (charList.size() > 1) {
            char lastChar = charList.get(charList.size() - 1);

            if (lastChar == '0') {
                charList.remove(charList.size() - 1);
            } else {
                addOne(charList);
            }

            ans += 1;
        }

        return ans;
    }

    private void addOne(List<Character> charList) {
        int i = charList.size() - 1;
        int carry = 1; 

        while (i >= 0 && carry == 1) {
            char c = charList.get(i);
            if (c == '0') {
                charList.set(i, '1');
                carry = 0;
            } else {
                charList.set(i, '0');
                
            }
            i--;
        }

        if (carry == 1) {
            charList.add(0, '1');
        }
    }
}